Route requests for missing static files using remote asset metadata

brandonpayton opened the pull request:
## What is this PR doing?

This PR:
- Includes a list of remote asset paths with minified WP builds
- Updates request handling to only handle missing files as static when they are explicitly listed as remote assets
(perhaps this should be split into multiple PRs)
- Updates request routing to behave more like a regular web server

Related to #1365 in that we need better information to judge whether to handle a request for a missing static file as PHP.

## What problem is it solving?

Prior to this PR, we could not easily tell whether we should request a missing static asset from the web or delegate the request to PHP.

## How is the problem addressed?

By including a list of remote asset paths with minified WP builds, we can judge which missing static file paths represent remote assets and which can be delegated to PHP.

## Testing Instructions

- CI tests
- Examine a minified WP zip and the contents of its playground-remote-asset-paths file
- Test manually with `npm run dev` and observe that Playground loads normally with no unexpected errors in the console

brandonpayton commented:
Apparently, a commit hook auto-committed a bunch of other changes that I had left outstanding. I will fix this and force push shortly.

brandonpayton commented:
> Apparently, a commit hook auto-committed a bunch of other changes that I had left outstanding. I will fix this and force push shortly.

No, I think that was me with an errant `git add -u`. I thought I'd seen some automated behavior like this in the past, but maybe I was wrong then as well. 🙃

brandonpayton commented:
The minified WP builds need to be updated as part of this PR. I had some trouble rebuilding them locally because the [command in the README](https://github.com/WordPress/wordpress-playground/blob/006cb0d5eab10d238388492b7a1fd332801a85d6/packages/playground/wordpress-builds/README.md#L7) didn't produce updated zips.

I ended up running the following commands to rebuilds the zips:
- `npx nx run playground-wordpress-builds:bundle-wordpress:major-and-beta`
- `npx nx run playground-wordpress-builds:bundle-wordpress:nightly`

This didn't update the zip for WP 6.1. It looks like we're still showing that as an option on playground.wordpress.net, but we only build the 3 previous major versions. Do we need to manually remove WP 6.1?

bgrgicak commented on packages/php-wasm/universal/src/lib/php-request-handler.ts:
```diff
@@ -492,6 +498,58 @@ export class PHPRequestHandler<PHP extends BasePHP> {
+	}
+
+	#seemsLikeAPHPFile(path: string) {
+		return path.endsWith('.php') || path.includes('.php/');
```
Should `.php/` also check for endsWith? This would prevent paths like `/test.php/index.js` from being executed.

```suggestion
		return path.endsWith('.php') || path.endsWith('.php/');
```

brandonpayton replied:
Yeah, that check is wrong, but I was leaving it for a follow up PR because it is fixing a different problem from this PR. Originally, I'd actually removed that check altogether, which is also the wrong thing to do.

I have a note to make a different PR for this and will make sure the `something.php/` case is tested.

brandonpayton replied:
> I have a note to make a different PR for this and will make sure the something.php/ case is tested.

Nevermind, we can go ahead and [refactor everything here](https://github.com/WordPress/wordpress-playground/pull/1417#discussion_r1604515501).

bgrgicak commented on packages/playground/remote/src/lib/worker-thread.ts:
```diff
@@ -184,6 +184,17 @@ try {
 			extractToPath: requestHandler.documentRoot,
 		});
 
+		// TODO: Move file to /internal?
+		// TODO: This feels like some kind of metadata we mentioned when discussing boot protocol
+		const remoteAssetListPath = `${requestHandler.documentRoot}/playground-remote-asset-paths`;
+		if (primaryPhp.fileExists(remoteAssetListPath)) {
+			const remoteAssetPaths = primaryPhp
+				.readFileAsText(remoteAssetListPath)
+				.split('\n');
+			requestHandler.addRemoteAssetPaths(remoteAssetPaths);
+			// TODO: Delete the listing file?
+		}
```
Yes to all TODOs. It's a custom file that's not part of WP and we should keep it in `internal`.

Deleting the file should help with preventing issues when the WP structure is updated, for example after a WP update, or version switch.

brandonpayton replied:
> Yes to all TODOs. It's a custom file that's not part of WP and we should keep it in internal

Sounds good!

bgrgicak commented on packages/playground/wordpress-builds/build/Dockerfile:
```diff
@@ -57,6 +57,11 @@ RUN cp -r wordpress wordpress-static && \
+# List assets that will be available to remotely so Playground can judge whether
+# to request a remote asset or delegate the request for a missing file to PHP.
+# TODO: What is a better name for this file?
+RUN find wordpress-static -type f | sed 's#^wordpress-static/##'> playground-remote-asset-paths
```
Maybe `wordpress-remote-asset-paths`?

```suggestion
RUN find wordpress-static -type f | sed 's#^wordpress-static/##'> wordpress-remote-asset-paths
```

bgrgicak replied:
But `playground-remote-asset-paths` also works.

brandonpayton replied:
🤔 The main point was to communicate that this is a special setup for minified WP builds, and it seemed a playground-specific thing to me. If I think about someone listing the files in one of the zips, I think it is clearer to see "playground-" than "wordpress-" when they ask "What is this thing?".

I was on the fence about this, but after talking it out, I think I'll leave it alone unless someone else feels strongly otherwise.

brandonpayton replied:
Heh, I save this comment and then think the opposite. Within playground, I would rather read `/internal/wordpress-remote-asset-paths`, so let's go with that.

adamziel replied:
Note this is only useful on the web and only for some WordPress builds but not for others. Also, the "active" WordPress build may change in runtime.

brandonpayton replied:
OK, so it sounds like it belongs with some kind of Wordpress-associated metadata.

I'll start by acting as if WP cannot be swapped out and address the request handling behavior and then address the possibility of WP switching, all within the scope of this PR.

brandonpayton replied:
> the "active" WordPress build may change in runtime.

@adamziel it looks like the web interface reloads the page every time the config changes. Where can the "active" WordPress build change anytime during runtime? Is this something that can happen with wp-now without reloading everything? (Planning to look further but wanted to get the question out there)

adamziel commented on packages/php-wasm/universal/src/lib/php-request-handler.ts:
```diff
@@ -313,7 +316,7 @@ export class PHPRequestHandler<PHP extends BasePHP> {
 		);
 		const fsPath = joinPaths(this.#DOCROOT, normalizedRequestedPath);
-		if (!seemsLikeAPHPRequestHandlerPath(fsPath)) {
+		if (!(await this.#seemsLikeAPHPRequestHandlerPath(fsPath))) {
```
At this point I'd refactor the routing logic completely. There's no more need for "seems" functions – we can resolve the URL to something like `{ "filePath": string, "isRemote": boolean }` with 100% certainty. Then we'd test whether `filePath` is a file with `.php` extension. A requested path like `/index.php/a/b/c` could mean different things, e.g. `I'm requesting a static file called "c" that's inside a directory tree like $DOCROOT/index.php/a/b/`, or `I'm requesting index.php and /a/b/c is my "nice URL"`. Let's do what a web server would do here.

bgrgicak replied:
If we do a full refactor let's also take this [use case into account. ](https://github.com/WordPress/wordpress-playground/pull/1187), I [started working on it,](https://github.com/WordPress/wordpress-playground/pull/1203) but lacked the metadata to complete it.

brandonpayton replied:
> Let's do what a web server would do here.

Right. Let's get it done. I'll work on that.

> If we do a full refactor let's also take this use case into account. 

Got it. Hopefully that will just naturally work after the refactor, but we can test it.

brandonpayton commented:
Here is a rough outline of how I think the handling should work:
```ts
let fsPath = joinPaths(this.#DOCROOT, normalizedRequestedPath);

// Simply handle PHP as PHP
if (fsPath.endsWith('.php')) {
	if (primaryPhp.isDir(fsPath)) {
		fsPath = joinPaths(fsPath, 'index.php');
	}

	if (primaryPhp.fileExists(fsPath)) {
		// TODO: Serve as PHP
	} else {
		// TODO: 404
	}
}

// Try as regular file
if (primaryPhp.fileExists(fsPath) && primaryPhp.isFile(fsPath)) {
	// TODO: Serve static file
} else if (this.#remoteAssetPaths.has(fsPath)) {
	// TODO: Request and serve static file
}

// Try as directory with default index.php
const localDefaultFile = joinPaths(fsPath, 'index.php');
if (
	primaryPhp.isDir(fsPath) &&
	primaryPhp.fileExists(localDefaultFile) &&
	primaryPhp.isFile(localDefaultFile)
) {
	// TODO: Serve as PHP
}

// TODO: Delegate request to WordPress /index.php, retaining query args and relaying original request URI
```

I still need to work through:
- how to actually dispatch PHP, delegating to WordPress `/index.php` while relaying original request URI and query args
- whether we want a different interface for identifying remote assets in PHPRequestHandler
- flagging remote assets for retrieval

adamziel commented:
It may require going left to right segment by segment until you reach a file, e.g. in `/wp-content/index.php/a/b` it would enter wp-content, find the index.php file, and stop resolving there. However, if index.php was a directory, it would go in and look for a node called `a`.

brandonpayton commented:
> It may require going left to right segment by segment until you reach a file, e.g. in /wp-content/index.php/a/b it would enter wp-content, find the index.php file, and stop resolving there. However, if index.php was a directory, it would go in and look for a node called a.

I have been working based on this bit of Nginx config which I do not believe does anything as complicated as that:
https://github.com/WordPress/Advanced-administration-handbook/blob/0f491e44c719abdf73989b278fc2bec8e4a4d03a/server/nginx.md#L162-L173
```nginx
    location / {
        # This is cool because no php is touched for static content.
        # include the "?$args" part so non-default permalinks doesn't break when using query string
        try_files $uri $uri/ /index.php?$args;
    }

    location ~ \.php$ {
        #NOTE: You should have "cgi.fix_pathinfo = 0;" in php.ini
        include fastcgi.conf;
        fastcgi_intercept_errors on;
        fastcgi_pass php;
    }
```

That is not to say we cannot do something more involved. But, as an example, WP Cloud doesn't directly run `/test.php` for the path `/test.php/a/b`. But it does delegate the request for `/test.php/a/b` to WordPress, which is what I believe [yesterday's sample logic](https://github.com/WordPress/wordpress-playground/pull/1417#issuecomment-2118535793) would do.

brandonpayton commented:
Hmm, unlike yesterday's sample logic, WP Cloud does delegate requests for missing PHP files to WordPress.

I don't think the WP.org's [example nginx config](https://github.com/WordPress/Advanced-administration-handbook/blob/0f491e44c719abdf73989b278fc2bec8e4a4d03a/server/nginx.md#L162-L173) would do that though for `.php` files. I plan to test this later.

brandonpayton commented:
I updated the routing logic in https://github.com/WordPress/wordpress-playground/pull/1417/commits/3f69fdcefb16a6bf57dab1b818badb78648cc6d6 to do roughly the following:
```
if dir URI:
    set PATH to default PHP file within that dir

if file has .php extension
    if file exists at PATH
        return PHP response
else
    if file exists at PATH
        return static file response
    if PATH is known remote asset
        return indication that path describes a remote asset

delegate all unhandled requests to WordPress /index.php
```

This behavior is close to [this bit of the WordPress.org nginx example](https://github.com/WordPress/Advanced-administration-handbook/blob/0f491e44c719abdf73989b278fc2bec8e4a4d03a/server/nginx.md#L162-L173)
```nginx
    location / {
        # This is cool because no php is touched for static content.
        # include the "?$args" part so non-default permalinks doesn't break when using query string
        try_files $uri $uri/ /index.php?$args;
    }

    location ~ \.php$ {
        #NOTE: You should have "cgi.fix_pathinfo = 0;" in php.ini
        include fastcgi.conf;
        fastcgi_intercept_errors on;
        fastcgi_pass php;
    }
```

But the implementation in this PR differs from the above config, unless I'm mistaken, because non-existent PHP files are delegated to WordPress. (In the config above, the `location ~\.php` block is given priority by nginx because it uses a regex, so `missing.php` should match that location and not match the location block where it tries paths and falls back to `/index.php`.) In my tests, WP Cloud does delegate missing PHP files to WordPress. This is more flexible behavior than failing outright, so I've adopted it here.

@adamziel I am not familiar with serving URLs like `/something/index.php/a/b` by running `/something/index.php`, at least not for WordPress. Is this needed?

So far, there are two main things left to do in this PR:
- Review existing tests and explicitly test all cases of this routing logic.
- Fix PR to handle switching WordPress versions.

brandonpayton commented:
One more note:
With this updated routing logic, all existing tests pass, and Playground seems to load without any issues.

brandonpayton commented:
Regarding sitemap.xml:
- Pretty permalinks must be enabled for things like sitemap.xml to be provided by WordPress
- Prior to these changes, requesting sitemap.xml through the top-level Playground web UI would 404 and show a blank page with an error logged to the console
- With these changes and pretty permalinks enabled, it delegates the request to WordPress and renders the response

![Screenshot 2024-05-21 at 1 56 14 PM](https://github.com/WordPress/wordpress-playground/assets/530877/65db6830-f002-416c-a04a-a47308cb3a98)

brandonpayton commented:
> Regarding sitemap.xml:
>
> Pretty permalinks must be enabled for things like sitemap.xml to be provided by WordPress

One thing I'm wondering:
AFAICT, the only way that WordPress actually returns a 404 status when handling a random, nonexistent request URI is to have pretty permalinks enabled. When those are enabled, WP responds with a pretty 404 page and an actual 404 status. 

When plain permalinks are selected and WordPress is handling a random, non-existent request URI, it responds with the home page and a 200, all without a redirect.

@adamziel @bgrgicak would it make sense to default WordPress within Playground to use pretty permalinks? Otherwise, all requests for non-existent files that we dedicate to WordPress will receive default homepage HTML and a 200 response status. It seems safer and clearer to actually have WordPress respond with 404s.

brandonpayton commented:
Given #1454 "Add missing trailing slash on landingPage wp-admin path", I think we should probably be responding with a redirect when a URI points to a directory without a trailing slash.

brandonpayton commented:
> Given https://github.com/WordPress/wordpress-playground/pull/1454 "Add missing trailing slash on landingPage wp-admin path", I think we should probably be responding with a redirect when a URI points to a directory without a trailing slash.

I added logic to redirect to add a trailing slash when an existing directory is requested without one. And the tests are now updated to cover the refactored routing logic.

The main thing remaining is to understand how the active WordPress build can change at runtime (as @adamziel [mentioned](https://github.com/WordPress/wordpress-playground/pull/1417#discussion_r1605190050)) and accommodate that in how we maintain the collection of known remote asset paths.

I guess the active WordPress build can change as simply as the filesystem can be changed over time. But are there other ways the entire WordPress version can change? Can multiple scopes use the same PHP request handler? These are the remaining questions for this PR.

cc @bgrgicak

bgrgicak commented:
> @adamziel @bgrgicak would it make sense to default WordPress within Playground to use pretty permalinks? Otherwise, all requests for non-existent files that we dedicate to WordPress will receive default homepage HTML and a 200 response status. It seems safer and clearer to actually have WordPress respond with 404s.

I would prefer to stick to the default WP behavior for permalink structures.

Does WP on Apache/NGINX also return 200?

brandonpayton commented:
> Does WP on Apache/NGINX also return 200?

Yes, this is my experience on WP Cloud as well when no permalink structure ("plain" permalinks) is selected.

> I would prefer to stick to the default WP behavior for permalink structures.

It turns out pretty permalinks are the default during WP install:
https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-admin/includes/upgrade.php#L471

That is good because features like sitemap.xml generation appear to depend on them being enabled.

Unfortunately, WP in Playground does not default to pretty permalinks, and I don't yet know why.

WP install does issue a request against the site to check whether pretty permalinks can be enabled by default:
https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-admin/includes/upgrade.php#L516-L522

I think that would require the changes in this PR to work. But it doesn't default to pretty permalinks with these changes either. Perhaps we're not running WP install yet or it runs differently than usual.

I made a note to file an issue for this if it isn't resolved in this PR.

brandonpayton commented on packages/playground/blueprints/src/lib/steps/login.spec.ts:
```diff
@@ -24,7 +24,7 @@ describe('Blueprint step installPlugin', () => {
 		await login(php, {});
 		const response = await requestHandler.request({
-			url: '/wp-admin',
+			url: '/wp-admin/',
```
This fixes the test because the handler now returns a redirect when the request URI identifies an existing dir without including a trailing slash.

brandonpayton commented on packages/playground/wordpress/src/index.ts:
```diff
@@ -49,14 +49,6 @@ export async function setupPlatformLevelMuPlugins(php: UniversalPHP) {
 	await php.writeFile(
 		'/internal/shared/mu-plugins/0-playground.php',
 		`<?php
-        // Redirect /wp-admin to /wp-admin/
```
This should no longer be necessary since the request handler now redirects to add a trailing slash for existing dirs.

brandonpayton commented:
I merged the latest from trunk and adapted these changes to work within the new bootWordPress() function. If I understand that function correctly, we get a new PHPRequestHandler each time it is called, so it seems like that should be enough to update the list of remote assets each time a new WordPress build is selected. If a build doesn't include a list of remote assets, we don't attempt to read it and assume all assets are included in the zip.

I believe this is ready for review.

bgrgicak commented on packages/php-wasm/universal/src/lib/php-request-handler.ts:
```diff
@@ -312,14 +315,64 @@ export class PHPRequestHandler<PHP extends BasePHP> {
+				);
+			}
+			const localDefaultPath = joinPaths(fsPath, 'index.php');
+			if (primaryPhp.isFile(localDefaultPath)) {
+				fsPath = localDefaultPath;
+			}
```
Should we check for `index.html` if `index.php` doesn't exist?

brandonpayton replied:
That seems like a good idea. I don't know of a reason we wouldn't want to serve index.html when there is no index.php.

brandonpayton replied:
Added in [f655598](https://github.com/WordPress/wordpress-playground/pull/1417/commits/f655598dc6d469bf02e78f1ff34da9a5cce054d9)

bgrgicak commented on packages/php-wasm/universal/src/lib/php-request-handler.ts:
```diff
@@ -312,14 +315,64 @@ export class PHPRequestHandler<PHP extends BasePHP> {
+				);
+			}
+		} else {
+			if (primaryPhp.isFile(fsPath)) {
```
NIT No need to nest this `if`.

brandonpayton replied:
Would you mind if I left it nested?

I bumped out the two blocks to just chain as else-ifs, but it turned out that leaving the else there helped my mind more clearly see the separation between how we handle PHP and static files. It is like:
```ts
if ( fileHasPhpExtension ) {
    // Here is what we do for PHP filenames
} else {
    // Here is what we do for everything else
}

// Everything not handled already is delegated to WP
```

bgrgicak replied:
> Would you mind if I left it nested?

All good, it just looked unusual to me, but I don't mind it/

brandonpayton commented:
> If I try to open an unexisting static file like `/sitemap.xml` this PR will redirect to `/sitemap.xml/` while trunk will return 404.
> 
> Ideally static files that don't exist should execute`/index.php`, but the path should stay unchanged.

This appears to be default WP behavior when plain permalinks are selected. The same occurs when testing from the WP codebase using `npm run env:start`.

adamziel commented on packages/php-wasm/universal/src/lib/php-request-handler.ts:
```diff
@@ -312,14 +315,70 @@ export class PHPRequestHandler<PHP extends BasePHP> {
+
+		// We can only satisfy requests for directories with a default file
+		// so let's first resolve to a default path when available.
+		if (primaryPhp.isDir(fsPath)) {
```
Let's add a more extensive note explaining how this is default WP behavior - it's not obvious this logic is the right one.

brandonpayton replied:
> Let's add a more extensive note explaining how this is default WP behavior - it's not obvious this logic is the right one.

Good point! Added this:
```ts
// Ensure directory URIs have a trailing slash. Otherwise,
// relative URIs in index.php or index.html files are relative
// to the next directory up.
//
// Example:
// For a request to "/wp-admin", the relative link "edit.php"
// resolves to "/edit.php" rather than "/wp-admin/edit.php".
//
// This is correct behavior for the browser:
// https://www.rfc-editor.org/rfc/rfc3986#section-5.2.3
//
// But the intent for `/wp-admin/index.php` is that its relative
// URIs are relative to `/wp-admin/`.
//
// In fact, WordPress also redirects like this when given a chance.
// - https://github.com/WordPress/wordpress-develop/blob/b6a3b9c7d1ce33cbeca6f95871a26c48141e524b/src/wp-includes/canonical.php#L696
// - https://github.com/WordPress/wordpress-develop/blob/b6a3b9c7d1ce33cbeca6f95871a26c48141e524b/src/wp-includes/canonical.php#L1036-L1045
// - https://github.com/WordPress/wordpress-develop/blob/b6a3b9c7d1ce33cbeca6f95871a26c48141e524b/src/wp-includes/link-template.php#L3558
```

adamziel reviewed (approved):
I left a nitpick that can
Be addressed in a follow-up PR alongside removing the /wp-admin redirect from platform mu-plugins. Otherwise this looks good. Thank you!
